### PR TITLE
Fix: Kits Defaults - Responsive values are not saved [ED-8987]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: modules/kit-elements-defaults/utils/settings-sanitizer.php [hunk 1]
```diff
@@ -94,15 +95,21 @@ public function remove_invalid_settings() {
...
+			) )
+			->map( function ( $value, $key ) use ( $available_settings ) {
 				if ( ! in_array( $key, static::SPECIAL_SETTINGS, true ) ) {
 					return $value;
 				}
 
-				return array_intersect_key( $value, $controls );
+				return ( new Collection( $value ) )
+					->only( $available_settings )
+					->all();
```
Comment (StyleShit) [inline modules/kit-elements-defaults/utils/settings-sanitizer.php:112]: I know some of the code isn't new, but maybe we can also add here some named variables/functions to make this part more readable.
I mean... I get what it does, but it takes some time to sink it. Especially when you don't know what it should do.

File: modules/kit-elements-defaults/utils/settings-sanitizer.php [hunk 2]
```diff
@@ -209,4 +216,32 @@ private function run_import_export_sanitize_process( Document $document, $proces
...
+	 * @return array
+	 */
+	private function get_avaliable_settings( $controls ) {
+		$control_keys = new Collection(
+			array_keys( $controls )
+		);
+
+		$optional_responsive_keys = new Collection(
+			Breakpoints_Manager::BREAKPOINT_KEYS
+		);
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/kit-elements-defaults/utils/settings-sanitizer.php:234]: It is a complex function that calculate the active breakpoints, and I want to keep it simple

File: modules/kit-elements-defaults/utils/settings-sanitizer.php [hunk 3]
```diff
@@ -209,4 +216,32 @@ private function run_import_export_sanitize_process( Document $document, $proces
...
 	}
+
+	/**
+	 * Get all the available settings of a specific element, including responsive settings.
+	 *
+	 * @param $controls
+	 *
+	 * @return array
+	 */
+	private function get_avaliable_settings( $controls ) {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/kit-elements-defaults/utils/settings-sanitizer.php:227]: Done

**Review by StyleShit — DISMISSED** [review]

File: modules/kit-elements-defaults/utils/settings-sanitizer.php [hunk 4]
```diff
@@ -209,4 +218,37 @@ private function run_import_export_sanitize_process( Document $document, $proces
...
+			Breakpoints_Manager::BREAKPOINT_KEY_TABLET_EXTRA,
+			Breakpoints_Manager::BREAKPOINT_KEY_LAPTOP,
+			Breakpoints_Manager::BREAKPOINT_KEY_WIDESCREEN,
+		];
+
+		$settings = [];
+
+		foreach ( $control_keys as $control_key ) {
+			// Add the responsive settings.
+			foreach ( $optional_responsive_keys as $responsive_key ) {
```
Comment (rotemee) [inline modules/kit-elements-defaults/utils/settings-sanitizer.php:245]: Not sure that I'm getting what's going on here, how do you know which controls are responsive?

File: modules/kit-elements-defaults/utils/settings-sanitizer.php [hunk 5]
```diff
@@ -95,16 +95,25 @@ public function remove_invalid_settings() {
...
+		}
+
+		$available_settings = $this->get_avaliable_settings( $controls );
 
-				return array_intersect_key( $value, $controls );
-			} )
-			->all();
+		foreach ( $this->pending_settings as $key => $value ) {
+			if ( in_array( $key, static::SPECIAL_SETTINGS, true ) ) {
+				// When the key is special settings like __dynamic__ or __globals__ it should run over the values.
```
Comment (rotemee) [inline modules/kit-elements-defaults/utils/settings-sanitizer.php:106]: Not sure that this comment is needed, consider adding a variable for the condition above:
$is_special_setting - like you had in the first place.

**Review by rotemee — CHANGES_REQUESTED** [review]

File: modules/kit-elements-defaults/utils/settings-sanitizer.php (hunk 4 above)
Comment (Nevoss) [reply modules/kit-elements-defaults/utils/settings-sanitizer.php:245]: I am assuming all the controls can be responsive, so I add the suffix for each control.

File: modules/kit-elements-defaults/utils/settings-sanitizer.php (hunk 5 above)
Comment (Nevoss) [reply modules/kit-elements-defaults/utils/settings-sanitizer.php:106]: Done

**Review by rotemee — DISMISSED** [review]

File: modules/kit-elements-defaults/utils/settings-sanitizer.php [hunk 6]
```diff
@@ -95,16 +95,25 @@ public function remove_invalid_settings() {
...
+			if ( $is_special_setting ) {
+				foreach ( $this->pending_settings[ $key ] as $special_setting_key => $special_setting_value ) {
+					if ( ! in_array( $special_setting_key, $available_settings, true ) ) {
+						unset( $this->pending_settings[ $key ][ $special_setting_key ] );
+					}
+				}
+			} elseif ( ! in_array( $key, $available_settings, true ) ) {
+				unset( $this->pending_settings[ $key ] );
+			}
+		}
```
Comment (StyleShit) [inline modules/kit-elements-defaults/utils/settings-sanitizer.php:116]: Maybe something like this?

```PHP
private function remove_unavailable_settings( $settings, $available_settings ) {
	foreach ( $settings as $key => $value ) {
		if ( ! in_array( $key, $available_settings, true ) ) {
			unset( $settings[ $key ] );
		}
	}
}
```

```PHP
$available_settings = $this->get_avaliable_settings( $controls );

$settings = $this->remove_unavailable_settings( 
	$this->pending_settings, 
	$available_settings + self::SPECIAL_SETTINGS  
);

foreach( self::SPECIAL_SETTINGS as $special_setting ) {
	$settings[ $special_setting ] = $this->remove_unavailable_settings( 
		$this->pending_settings, 
		$available_settings 
	);
}

return $settings;
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

**Review by StyleShit — APPROVED** [review]

**Pull request merged**